src/annotator/util/test/buckets-test.js

```diff
@@ -0,0 +1,276 @@
+      assert.isNull(findClosestOffscreenAnchor(fakeAnchors, 'down'));
+    });
+  });
+
+  describe('constructPositionPoints', () => {
+    let fakeAnchors;
+    let stubbedInnerHeight;
+
+    beforeEach(() => {
+      const fakeAnchor = fakeAnchorFactory();
+      fakeAnchors = [
+        fakeAnchor(), // top: 1
+        fakeAnchor(), // top: 101
+        fakeAnchor(), // top: 201
+        fakeAnchor(), // top: 301
+        fakeAnchor(), // top: 401
+        fakeAnchor(), // top: 501
+      ];
+      stubbedInnerHeight = sinon.stub(window, 'innerHeight').value(410);
+    });
+
+    afterEach(() => {
+      stubbedInnerHeight.restore();
+    });
+
+    it('returns an Array of anchors that are offscreen above', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      assert.deepEqual(positionPoints.above, [fakeAnchors[0], fakeAnchors[1]]);
+    });
+
+    it('returns an Array of anchors that are offscreen below', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      assert.deepEqual(positionPoints.below, [fakeAnchors[4], fakeAnchors[5]]);
+    });
+
+    it('does not return duplicate anchors', () => {
+      const positionPoints = constructPositionPoints([
+        fakeAnchors[0],
+        fakeAnchors[0],
+        fakeAnchors[5],
+        fakeAnchors[5],
+      ]);
+
+      assert.deepEqual(positionPoints.above, [fakeAnchors[0]]);
+      assert.deepEqual(positionPoints.below, [fakeAnchors[5]]);
+    });
+
+    it('returns an Array of position points for on-screen anchors', () => {
+      const positionPoints = constructPositionPoints(fakeAnchors);
+
+      // It should return two "point" positions for each on-screen anchor,
+      // one representing the top of the anchor's highlight box, one representing
+      // the bottom position
+      assert.equal(positionPoints.points.length, 4);
+      // The top position of the first on-screen anchor
+      assert.deepEqual(positionPoints.points[0], [201, 1, fakeAnchors[2]]);
+      // The bottom position of the first on-screen anchor
+      assert.deepEqual(positionPoints.points[1], [251, -1, fakeAnchors[2]]);
+      // The top position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[2], [301, 1, fakeAnchors[3]]);
+      // The bottom position of the second on-screen anchor
+      assert.deepEqual(positionPoints.points[3], [351, -1, fakeAnchors[3]]);
+    });
+
+    it('sorts on-screen points based on position primarily, type secondarily', () => {
+      fakeGetBoundingClientRect.callsFake(() => {
+        return {
+          top: 250,
+          bottom: 250,
+        };
+      });
+      const positionPoints = constructPositionPoints(fakeAnchors);
+      for (let i = 0; i < fakeAnchors.length; i++) {
+        // The bottom position for all of the fake anchors is the same, so
+        // those points will all be at the top of the list
+        assert.equal(positionPoints.points[i][2], fakeAnchors[i]);
```

Comment:
Reading this again I find it counter-intuitive that "top" points have a type of 1 and "bottom" points have a type of -1 and that the sorting is done such that "top" points appear after "bottom" points. I see this was the case in the previous code though.

Reply:
I also think it's technically _fragile_ as the whole thing would break down if it saw a bottom point before a top point for a given anchor. That "shouldn't happen", but it still feels funky.